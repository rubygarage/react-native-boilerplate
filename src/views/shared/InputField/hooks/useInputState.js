import { useState } from 'react';
import { call, path, pathOr } from 'ramda';

function useInputState(props) {
  const {
    field: { name },
    form: { handleBlur, errors },
  } = props;

  const [isFocused, setIsFocused] = useState(false);
  const [actionWidth, setActionWidth] = useState(0);

  /**
   * On Action required.
   */
  const onActionRendered = (event) => {
    const { width } = event.nativeEvent.layout;

    setActionWidth(width);
  };

  /**
   * On focus handler.
   */
  const onFocus = () => {
    setIsFocused(true);
  };

  /**
   * On Blur handler
   */
  const onBlur = (event) => {
    call(handleBlur(name), event);
    setIsFocused(false);
  };

  /**
   * Error data
   */
  const getErrorData = () => {
    const errorId = path([name, 'id'], errors);
    const errorValues = path([name, 'values'], errors);
    const errorIcon = pathOr('alert', [name, 'icon'], errors);
    return { errorId, errorValues, errorIcon };
  };

  return {
    isFocused,
    actionWidth,
    onActionRendered,
    onFocus,
    onBlur,
    getErrorData,
    setIsFocused,
    setActionWidth,
  };
}

export default useInputState;
