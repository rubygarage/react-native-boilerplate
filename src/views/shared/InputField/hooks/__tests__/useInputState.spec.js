import { renderHook, act } from '@testing-library/react-hooks';
import { call } from 'ramda';

import useInputState from '../useInputState';

jest.mock('ramda', () => ({
  ...jest.requireActual('ramda'),
  call: jest.fn(),
}));

describe('useInputState custom hook', () => {
  const props = {
    field: { name: 'test' },
    form: {
      handleBlur: jest.fn(),
      errors: {},
    },
  };

  it('matches snapshot', () => {
    const { result } = renderHook(() => useInputState(props));

    expect(result.current).toMatchSnapshot();
  });

  it('checks onActionRendered method', () => {
    const { result } = renderHook(() => useInputState(props));

    const event = {
      nativeEvent: {
        layout: {
          width: 444,
        },
      },
    };

    act(() => {
      result.current.onActionRendered(event);
    });

    expect(result.current.actionWidth).toBe(444);
  });

  it('checks onFocus method', () => {
    const { result } = renderHook(() => useInputState(props));

    act(() => {
      result.current.onFocus();
    });

    expect(result.current.isFocused).toBe(true);
  });

  it('checks onBlur method', () => {
    const event = {};

    const { result } = renderHook(() => useInputState(props));

    act(() => {
      result.current.onBlur(event);
    });

    expect(result.current.isFocused).toBe(false);
    expect(call).toHaveBeenCalledWith(props.form.handleBlur('test'), event);
  });

  describe('checks getErrorData method', () => {
    it('should return undefined', () => {
      const { result } = renderHook(() => useInputState(props));

      let errors = null;

      act(() => {
        errors = result.current.getErrorData();
      });

      expect(errors).toEqual({
        errorId: undefined,
        errorValues: undefined,
        errorIcon: 'alert',
      });
    });
  });

  it('should return expected values', () => {
    const { result } = renderHook(() => useInputState({
      ...props,
      form: {
        handleBlur: jest.fn(),
        errors: {
          test: {
            id: 'test_id',
            values: 'test_values',
            icon: 'test_icon',
          },
        },
      },
    }));

    let errors = null;

    act(() => {
      errors = result.current.getErrorData();
    });

    expect(errors).toEqual({
      errorId: 'test_id',
      errorValues: 'test_values',
      errorIcon: 'test_icon',
    });
  });
});
