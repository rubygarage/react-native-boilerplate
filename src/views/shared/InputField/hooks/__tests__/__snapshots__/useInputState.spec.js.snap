// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useInputState custom hook matches snapshot 1`] = `
Object {
  "actionWidth": 0,
  "getErrorData": [Function],
  "isFocused": false,
  "onActionRendered": [Function],
  "onBlur": [Function],
  "onFocus": [Function],
  "setActionWidth": [Function],
  "setIsFocused": [Function],
}
`;
